### PR TITLE
[FLINK-33199][tests] Use fully qualified class names instead of Class objects in ArchitectureTests

### PR DESCRIPTION
**Pull request by afedulov**
## What is the purpose of the change

Currently architecture tests rely on importing such classes as `MiniClusterExtension`. This introduces a production scope dependency on `flink-test-utils` which in turn depends on `flink-streaming-java`. This is problematic because adding architecture tests to any direct or transitive dependency of `flink-streaming-java` creates a dependency cycle.
Example: https://github.com/apache/flink/pull/22850#discussion_r1243343382
In general, since architecture tests are supposed to be used freely in any submodule, it is desirable to reduce its dependency surface as much as possible to prevent such cycles. 

## Brief change log
  - Modifies methods to using  fully qualified type names checks instead of Class objects
  - Removes production maven dependencies on `flink-test-utils`,  `flink-test-utils-junit` and `flink-connector-test-utils`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation
  - Does this pull request introduce a new feature? (yes / **no**)

### CODE REVIEW

File: flink-architecture-tests/flink-architecture-tests-base/src/main/java/org/apache/flink/architecture/common/Predicates.java [hunk 1]
```diff
@@ -157,6 +166,35 @@ public static <T> DescribedPredicate<T> exactlyOneOf(
...
+    /**
+     * Extracts the class name from the given fully qualified class name.
+     *
+     * <p>Example:
+     *
+     * <pre>
+     *     getClassFromFqName("com.example.MyClass");  // Returns: "MyClass"
+     * </pre>
+     */
+    public static String getClassSimpleNameFromFqName(String fqClassName) {
```
Comment (zentol) [inline flink-architecture-tests/flink-architecture-tests-base/src/main/java/org/apache/flink/architecture/common/Predicates.java:178]: One thing to be careful of is what kind of fqcn you are using.

Canonical names can for example contain a `$` before the simple name if you are dealing with an inner class. Or its the other way around that getName returns it; this always trips me up.

File: flink-architecture-tests/flink-architecture-tests-base/src/main/java/org/apache/flink/architecture/common/Predicates.java [hunk 2]
```diff
@@ -62,40 +68,42 @@ public static DescribedPredicate<JavaClass> containAnyFieldsInClassHierarchyThat
     }
 
     /**
-     * Tests that the given field is {@code public static} and of the given type {@code clazz} .
+     * Tests that the given field is {@code public static} and has the fully qualified type name of
+     * {@code fqClassName}.
      *
      * <p>Attention: changing the description will add a rule into the stored.rules.
      */
-    public static DescribedPredicate<JavaField> arePublicStaticOfType(Class<?> clazz) {
```
Comment (zentol) [inline flink-architecture-tests/flink-architecture-tests-base/src/main/java/org/apache/flink/architecture/common/Predicates.java:69]: Do we know whether any of these methods are also used by connector repos? Wondering about compile error when working against newer Flink versions.

Comment (afedulov) [reply flink-architecture-tests/flink-architecture-tests-base/src/main/java/org/apache/flink/architecture/common/Predicates.java:69]: Good point! I ran a script to search for the following methods
```
areFieldOfType
arePublicFinalOfType
arePublicStaticOfType
arePublicStaticFinalOfType
arePublicFinalOfTypeWithAnnotation
arePublicStaticFinalOfTypeWithAnnotation
areStaticFinalOfTypeWithAnnotation
```
across these repositories
```
apache/flink-connector-kafka
apache/flink-connector-cassandra
apache/flink-connector-elasticsearch
apache/flink-connector-rabbitmq
apache/flink-connector-google-cloud-pubsub
apache/flink-connector-pulsar
apache/flink-connector-jdbc
apache/flink-connector-hbase
apache/flink-connector-hive
apache/flink-connector-aws
apache/flink-connector-firehose
apache/flink-connector-kinesis
apache/flink-connector-dynamodb
```
and did not find any usage.

It seems that the question "Does this change in Flink break any connector?" will be a recurring topic with the externalized repos. Having some automated checks is tricky, but feasible:
1) get all the method signature changes in the PR
2) search across the connector repos for usage of the respective classes + methods
The experience I made is that it won't work with GitHub API due to pretty heavy throttling on their side. What do, will  checking out all the connector repos in the main Flink CI run work for us?

File: flink-architecture-tests/flink-architecture-tests-base/src/main/java/org/apache/flink/architecture/common/Predicates.java (hunk 1 above)
Comment (afedulov) [reply flink-architecture-tests/flink-architecture-tests-base/src/main/java/org/apache/flink/architecture/common/Predicates.java:178]: Thanks for pointing this out. I changed the approach to handle both cases and added test coverage for this method:
https://github.com/apache/flink/pull/23491/commits/fbea7c9382bd9d3f6a7d80c81e05e8491a137612

**Comment by afedulov** [comment]
@zentol thanks, Chesnay! I addressed your comments.

File: flink-architecture-tests/flink-architecture-tests-base/src/main/java/org/apache/flink/architecture/common/Predicates.java (hunk 2 above)
Comment (zentol) [reply flink-architecture-tests/flink-architecture-tests-base/src/main/java/org/apache/flink/architecture/common/Predicates.java:69]: > What do think, could checking out all the connector repos in the main Flink CI run work for us?

I think we just to clarify what is part of the user (connector?) facing API, annotate them accordingly, and them treat them as other public APIs.

File: flink-architecture-tests/flink-architecture-tests-base/src/test/java/PredicatesTest.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import org.junit.jupiter.api.Test;
+
+import static org.apache.flink.architecture.common.Predicates.getClassSimpleNameFromFqName;
+import static org.assertj.core.api.Assertions.*;
```
Comment (bvarghese1) [inline flink-architecture-tests/flink-architecture-tests-base/src/test/java/PredicatesTest.java:22]: @afedulov I think this `.*` is causing the compile step to fail with a Checkstyle error

Comment (bvarghese1) [reply flink-architecture-tests/flink-architecture-tests-base/src/test/java/PredicatesTest.java:22]: https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=53704&view=logs&j=52b61abe-a3cc-5bde-cc35-1bbe89bb7df5&t=54421a62-0c80-5aad-3319-094ff69180bb

Comment (afedulov) [reply flink-architecture-tests/flink-architecture-tests-base/src/test/java/PredicatesTest.java:22]: @bvarghese1 indeed

Comment (afedulov) [reply flink-architecture-tests/flink-architecture-tests-base/src/test/java/PredicatesTest.java:22]: @zentol please merge this hotfix https://github.com/apache/flink/pull/23522/files

**Pull request merged**